src/main/java/core/basesyntax/SalaryInfo.java

```diff
@@ -39,8 +43,72 @@
      * Андрей - 600
      * София - 900</p>
      */
-    public String getSalaryInfo(String[] names, String[] data, String dateFrom, String dateTo)
-            throws Exception {
-        return null;
+
+    public String getSalaryInfo(String[] names, String[] data,
+                                String dateFrom, String dateTo) throws Exception {
+        checkInputDates(dateFrom, dateTo);
+
+        Worker[] workers = new Worker[names.length];
+
+        for (int i = 0; i < names.length; i++) {
+            Worker worker = new Worker(names[i]);
+            workers[i] = worker;
+        }
+
+        for (String singleData : data) {
+            DataLineInfo dataLineInfo = new DataLineInfo(singleData);
+            if (isNeedDate(dateFrom, dateTo, dataLineInfo.date)) {
+                for (Worker worker : workers) {
+                    if (worker.name.equals(dataLineInfo.name)) {
+                        worker.salary += dataLineInfo.hours * dataLineInfo.payment;
+                    }
+                }
+            }
+        }
+
+        String report = makeReport(dateFrom, dateTo, workers);
+        return report;
+    }
+
+    private String makeReport(String dateFrom, String dateTo, Worker[] workers) {
+        StringBuilder result = new StringBuilder();
+
+        result.append("Отчёт за период ")
+                .append(dateFrom)
+                .append(" - ")
+                .append(dateTo)
+                .append("\n");
+
+        for (Worker worker : workers) {
+            result.append(worker.name).append(" - ").append(worker.salary).append("\n");
+            ;
+        }
+        result.append("\n");
+
+        return result.toString().replace("\n\n", "");
```

Comment:
why do you add \n twice and then remove it?

Reply:
because test failed with '\n' in the end of report